toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagecustomizerapi"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagegen/diskutils"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/file"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/shell"
+)
+
+func CustomizeImageWithConfigFile(buildDir string, configFile string, imageFile string,
+	outputImageFile string, outputImageFormat string,
+) error {
+	var err error
+
+	var config imagecustomizerapi.SystemConfig
+	err = imagecustomizerapi.UnmarshalYamlFile(configFile, &config)
+	if err != nil {
+		return err
+	}
+
+	baseConfigPath, _ := filepath.Split(configFile)
+
+	err = CustomizeImage(buildDir, baseConfigPath, &config, imageFile, outputImageFile, outputImageFormat)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func CustomizeImage(buildDir string, baseConfigPath string, config *imagecustomizerapi.SystemConfig, imageFile string,
+	outputImageFile string, outputImageFormat string,
+) error {
+	var err error
+
+	// Validate 'outputImageFormat' value.
+	qemuOutputImageFormat, err := toQemuImageFormat(outputImageFormat)
+	if err != nil {
+		return err
+	}
+
+	// Normalize 'buildDir' path.
+	buildDirAbs, err := filepath.Abs(buildDir)
+	if err != nil {
+		return err
+	}
+
+	// Create 'buildDir' directory.
+	err = os.MkdirAll(buildDirAbs, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	// Validate config.
+	err = validateConfig(baseConfigPath, config)
+	if err != nil {
+		return fmt.Errorf("invalid image config: %w", err)
+	}
+
+	// Convert image file to raw format, so that a kernel loop device can be used to make changes to the image.
+	buildImageFile := filepath.Join(buildDirAbs, "image.raw")
+
+	_, _, err = shell.Execute("qemu-img", "convert", "-O", "raw", imageFile, buildImageFile)
```

Comment:
Do we need to check the current image format to see if this step is necessary? or do we only accept vhd as input?

Reply:
If the file format is the same, then `qemu-img convert` will just copy the file. This is actually what I want to ensure that the original file remains unmodified.